hypothesis/client
# Add option to export annotations in markdown

This PR adds a new annotations export format, Markdown.

### Some decisions

* I initially considered using the HTML format and parsing it to Markdown for simplicity, but it presented some problems:
  * We don't have a dependency to do that. The most used and up-to-date one seems to be [turndown](https://www.npmjs.com/package/turndown), however, in order to render tables, we needed a plugin to add GitHub markup flavor, and the only one is very outdated (6 years since last release) and lacks TS definitions, even via "definitely typed" project, and ultimately, it didn't work, leaving table markup un parsed.
    I looked for other options, but none of them seemed great.
  * The HTML structure seemed a bit too complex for markdown. An result which is between the plain text and the HTML ones seems more sensible.
 
  This made me decide to manually render the markdown, as we do with other formats.
* As mentioned above, and in order to have something closer to the text format, I decided to use bullet points for the info on every annotations and for the summary, as opposed to a table as we do in HTML.
* Since annotations text contains markdown, I decided to keep it like that, as it is consistent with the export format. However, I wrapped in a `pre`, as a way to "escape" it and avoid breaking the overall format.
* Quotes are wrapped in a `blockote`, but trimming every line, otherwise their meaning changes and can potentially render differently depending on where the markdown is used.
  For example, GitHub interprets extra indentation as a `pre` block.

### Testing steps

1. Open the export annotations panel.
2. Select `Markdown` from the format dropdown.
3. Click "Export" -> You should get a file with `md` extension containing the markdown.
4. Click "Copy to clipboard" -> It should copy the same markdown to the clipboard.
5. Paste the result in a new `GitHub` issue and click "Preview" -> It should properly render everything.

### Todo

- [ ] Come up with a description for Markdown.
- [ ] Decide the right short title. Currently used just `MD` as `Markdown` does not fit the size of the dropdown, but maybe we want to increase that size instead.
